### PR TITLE
Tested Database Functions

### PR DESCRIPTION
**Pull request by MalikG15**

### CODE REVIEW

File: src/codeu/chat/DerbyStore.java [hunk 1]
```diff
@@ -35,23 +35,24 @@
 	private final String messageTableName = "message";
 	private final String chatParticipantsTableName = "chatParticipants";
 	
-	private final String getUserInfo = "SELECT * FROM " + userTableName +  " WHERE id = ?";
-	private final String getConversationInfo = "SELECT * FROM " + userTableName +  " WHERE id = ?";
-	private final String getUsersInvolved = "SELCT * FROM " + chatParticipantsTableName + " WHERE id = ?" ;
+	private final String getUserInfo = "SELECT * FROM " + userTableName +  " WHERE userid = ?";
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:38]: nit: This is okay for our purposes, but with more complicated tables you probably want to select specific columns instead of selecting *.

File: src/codeu/chat/DerbyStore.java [hunk 2]
```diff
@@ -35,23 +35,24 @@
...
+	private final String addMessageInfo = "INSERT INTO " + messageTableName + " (messageid, previous, creation, author, content, nextMessage) " + 
 	"values (?, ?, ?, ?, ?, ?)";
 	private final String addChatParticipantsInfo = "INSERT INTO " + chatParticipantsTableName + " (conversationid, userid) values (?, ?)";
 
 	
-	private final String updateLastMessageById = "UPDATE " + messageTableName + " SET nextMessage = ? WHERE id = ?";
-	private final String updateConversationById = "UPDATE " + conversationTableName + " SET firstMessage = ?, lastMessage = ? WHERE id = ?";
+	private final String updateNextMessageById = "UPDATE " + messageTableName + " SET nextMessage = ? WHERE messageid = ?";
+	private final String updateConversationById = "UPDATE " + conversationTableName + " SET firstMessage = ?, lastMessage = ? WHERE conversationid = ?";
+	private final String getMessageById = "SELECT * FROM " + messageTableName + " WHERE messageid = ?";
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:55]: (commenting here because I can't comment on unchanged lines)

- I'd add some javadoc comments on the public methods and the constructor.
- You're still hardcoding one of the fields in the `addUser()` function.

File: src/codeu/chat/DerbyStore.java [hunk 3]
```diff
@@ -81,18 +82,18 @@ public DerbyStore() {
 			stmt = conn.createStatement();
 			
 			// Create the chat user table
-			stmt.execute("CREATE TABLE " + userTableName + "(id varchar(255), name varchar(255), password varchar(255), creation BIGINT)");
+			stmt.execute("CREATE TABLE " + userTableName + " (userid varchar(255), name varchar(255), password varchar(255), creation BIGINT, PRIMARY KEY (userid))");
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:85]: nit: maaaybe some comments explaining what the userid (and other primary keys below) is would be helpful? Also, is name guaranteed to be unique?

File: src/codeu/chat/DerbyStore.java [hunk 4]
```diff
@@ -105,36 +106,43 @@ public DerbyStore() {
 	@Override
 	public User getUser(Uuid userid) throws SQLException, IOException {
 		PreparedStatement getUser = conn.prepareStatement(getUserInfo);
-		getUser.setString(1, userid.toString());
+		getUser.setString(1, removeCharsInUuid(userid.toString()));
 		ResultSet user = getUser.executeQuery();
 		
-		return new User(Uuid.parse(user.getString(1)), user.getString(2), Time.fromMs(user.getLong(2)));
+		if (user.next()) {
+			return new User(Uuid.parse(user.getString(1)), user.getString(2), Time.fromMs(user.getLong(4)));
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java]: nit: I'd probably use column names (stored in variables) instead of indexes. Same thing for indexes in other places.

File: src/codeu/chat/DerbyStore.java [hunk 5]
```diff
@@ -202,11 +210,10 @@ public void addConversation(Conversation c) throws SQLException {
 		addConversation.setString(5, removeCharsInUuid(c.firstMessage.toString()));
 		addConversation.setString(6, removeCharsInUuid(c.lastMessage.toString()));
 		addConversation.executeUpdate();
-		
 	}
 	
 	@Override
-	public void updateConversation(Conversation c) throws SQLException {
+	public void updateConversationMessages(Conversation c) throws SQLException {
```
Comment (KevinWorkman) [inline src/codeu/chat/DerbyStore.java:216]: Some javadoc (maybe on the interface) explaining when these functions should be called would be helpful.

File: src/codeu/chat/common/DerbyDatabaseInteractions.java [hunk 6]
```diff
@@ -27,13 +27,15 @@
 	
 	public void addConversation(Conversation c) throws SQLException;
 	
-	public void updateConversation(Conversation c) throws SQLException;
+	public void updateConversationMessages(Conversation c) throws SQLException;
```
Comment (KevinWorkman) [inline src/codeu/chat/common/DerbyDatabaseInteractions.java:30]: Some javadoc comments explaining when these functions should be called would be helpful.

File: src/codeu/chat/server/Model.java [hunk 7]
```diff
@@ -182,7 +182,7 @@ public void updateMessageLink(Message message) {
 	  // EDIT - Malik Graham
 	  // Updating the next message for the last message
       try {
-    	  ds.updateLastMessage(message);
+    	  ds.updateNextMessage(message);
```
Comment (KevinWorkman) [inline src/codeu/chat/server/Model.java:185]: Why would these functions throw an exception? Could we handle them a little better than just printing the stack trace?

File: test/codeu/chat/server/DerbyStoreTest.java [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+import org.junit.Before;
+
+import codeu.chat.DerbyStore;
+import codeu.chat.common.User;
+import codeu.chat.util.Uuid;
+import codeu.chat.common.BasicController;
+import codeu.chat.common.Conversation;
+import codeu.chat.common.Message;
+
+public class DerbyStoreTest {
```
Comment (KevinWorkman) [inline test/codeu/chat/server/DerbyStoreTest.java:14]: Hooray! So happy to see a test class.

File: test/codeu/chat/server/DerbyStoreTest.java [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+	
+	private DerbyStore ds;
+	private Model model;
+	private BasicController controller;
+		
+	@Before
+	public void doBefore() {
+		ds = new DerbyStore();
+		model = new Model();
+	    controller = new Controller(Uuid.NULL, model);
```
Comment (KevinWorkman) [inline test/codeu/chat/server/DerbyStoreTest.java]: nit: indentation. If you're using Eclipse, autoformatting should take care of this. I press Ctrl+A, Ctrl+I, Ctrl+S (select all, autoformat, save) pretty much every time I write a line of code. :smile_cat:

File: test/codeu/chat/server/DerbyStoreTest.java [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+	@Test
+	public void checkUserNameExists() {
+		final User user = controller.newUser("user");
+		boolean userNameExists = false;
+		
+		try {
+			userNameExists = ds.checkUsernameExists(user.name);
+		}
+		catch (Exception ex) {
+
```
Comment (KevinWorkman) [inline test/codeu/chat/server/DerbyStoreTest.java]: Empty catch blocks are usually a bad sign. It's not the end of the world here because the test will fail anyway, but it's probably a good idea to at least toss a `printStackTrace()` in here.

File: test/codeu/chat/server/DerbyStoreTest.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+			
+		}
+		
+		assertTrue("Check to see if a username already exists",
+				userNameExists);
+	}
+	
+	@Test
+	public void addAndGetUser() {
+		 final User user = controller.newUser("user");
```
Comment (KevinWorkman) [inline test/codeu/chat/server/DerbyStoreTest.java]: nit: I would rename these to something like `savedUser` and `loadedUser`.

File: test/codeu/chat/server/DerbyStoreTest.java [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+		 User testUser = null;
+		 try {
+			//ds.addUser(user);
+			 testUser = ds.getUser(user.id);
```
Comment (KevinWorkman) [inline test/codeu/chat/server/DerbyStoreTest.java]: From a design perspective, it seems a little weird to go to the `DerbyStore` for some stuff and the `Controller` for other stuff. I would expect the `Controller` to sit on top of the `DerbyStore`, so you wouldn't go directly to the `DerbyStore` for anything?

Then I'd have two separate test classes: one for the `DerbyStore` that tests low-level saving and loading (and stuff like primary keys), and another one for `Controller` that uses a mocked `DerbyStore` to test the controller logic.

File: test/codeu/chat/server/DerbyStoreTest.java [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+		catch (Exception ex) {
+			ex.printStackTrace();
+		}
+		
+		assertTrue(
+				 "Check that messages can be retrieved and have the same id",
+				 Uuid.equals(message.next, nextMessage.id));
+		
+	}
+
```
Comment (KevinWorkman) [inline test/codeu/chat/server/DerbyStoreTest.java]: So far these test how things should work, but you'd probably want to add a few tests that make sure things break correctly, too. What happens if you try to add a row with the same primary key as an existing row? Try to think of as many corner cases as you can.

File: test/codeu/chat/server/DerbyStoreTest.java [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+				 "Check that messages can be retrieved and have the same id",
+				 Uuid.equals(message.next, nextMessage.id));
+		
+	}
+	
+	
+	// TODO: Add a way to get messages based on conversations and test it
+	// 		 Add a method to update the users involved for a conversation
+	
+
```
Comment (KevinWorkman) [inline test/codeu/chat/server/DerbyStoreTest.java]: You probably want to use a different database for testing than you do for your actual program, and you probably want to delete the test database after each test.

**Review by KevinWorkman — COMMENTED** [review]
Overall nice work. Few comments and stuff to think about. Awesome to see the tests coming together.

I'm a little curious about the design (you go to the controller for some stuff and directly to the database for other stuff), and I think you probably want a separate database for testing, but in general this is great progress.

File: test/codeu/chat/server/DerbyStoreTest.java (hunk 12 above)
Comment (MalikG15) [reply test/codeu/chat/server/DerbyStoreTest.java]: I'm using the controller object to create a new conversation, message, or user. The controller has functionality already that saves the object in a Store<Uuid, Object> variable, and I've added code that also puts it into the database. So, every time I make a user object for example it is saved. I'm using DerbyStore to pull that object once its saved.

File: src/codeu/chat/server/Model.java (hunk 7 above)
Comment (MalikG15) [reply src/codeu/chat/server/Model.java:185]: It throws an exception because we're making a call to the database, and whenever we do that a SQL Exception could possibly be thrown.

File: src/codeu/chat/DerbyStore.java (hunk 3 above)
Comment (MalikG15) [reply src/codeu/chat/DerbyStore.java:85]: Name, for now, is guaranteed to be unique. That will be their username (so I should change the name of that column to username). Could I possibly have two primary keys?

File: src/codeu/chat/DerbyStore.java (hunk 2 above)
Comment (MalikG15) [reply src/codeu/chat/DerbyStore.java:55]: We don't have a way yet of producing a password yet so I'm hardcoding it.

File: src/codeu/chat/DerbyStore.java (hunk 1 above)
Comment (MalikG15) [reply src/codeu/chat/DerbyStore.java:38]: I want all of the users' info to create a complete User object from the information that was saved.

Comment (KevinWorkman) [reply src/codeu/chat/DerbyStore.java:38]: Yeah that's why it's okay for our purposes. But in "real life" you'd probably want to be specific about exactly what columns you're getting, even if you're getting them all. Not saying you have to change it here, just something to think about.

File: src/codeu/chat/DerbyStore.java (hunk 2 above)
Comment (KevinWorkman) [reply src/codeu/chat/DerbyStore.java:55]: Understood.

File: src/codeu/chat/DerbyStore.java (hunk 3 above)
Comment (KevinWorkman) [reply src/codeu/chat/DerbyStore.java:85]: Yep, you can absolutely have two primary keys. (You can also have a primary key that's based on multiple columns, but that's not what you want here.) Then you'd probably want to add more robust handling of the case where a user tries to register with a name that's already used.

File: src/codeu/chat/server/Model.java (hunk 7 above)
Comment (KevinWorkman) [reply src/codeu/chat/server/Model.java:185]: I guess the question is, will that exception ever actually happen, since we're running sql that you wrote (as opposed to sql that a user wrote)? Maybe you can do the exception catching at a different level (inside `DerbyStore`), so calling code doesn't have to worry about it?

File: test/codeu/chat/server/DerbyStoreTest.java (hunk 12 above)
Comment (KevinWorkman) [reply test/codeu/chat/server/DerbyStoreTest.java]: Hmm okay, whatever you think is best is fine with me. I would just maybe expect to see two different levels: the database level that deals with low-level data access, and the controller level that sits on top of that and contains the logic. Right now they're a little mixed together.

Not saying you have to fix that right now, most of these comments are just stuff for you to think about. There isn't any single correct way to do most of this stuff!
